### PR TITLE
OSCI driver builds

### PR DESCRIPTION
**Pull request by Molter73**
## Description

This PR holds all the necessary changes for OSCI to build kernel drivers, PR openshift/release#29021 holds the OSCI configuration side that goes with these changes.

## Checklist
- [ ] Investigated and inspected CI test results
~- [ ] Updated documentation accordingly~

**Automated testing**
~- [ ] Added unit tests~
~- [ ] Added integration tests~
~- [ ] Added regression tests~

If any of these don't apply, please comment below.

## Testing Performed

Been running rehearses on the OSCI PR, testing various combinations of builds and seems to be mostly working. Any further problems that might show up after merge will be addressed in follow up PRs.

### CODE REVIEW

File: .circleci/kernel-module-build-failures-check/20-test-for-build-failures.sh
Comment (JoukoVirtanen) [inline .circleci/kernel-module-build-failures-check/20-test-for-build-failures.sh:8]: Is there a difference in performance between this and using "find"? Perhaps find isn't installed. Probably not worth changing. I leave it up to you.

File: .openshift-ci/drivers/BLOCKLIST
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/BLOCKLIST:9]: [Nit] Consider adding a blank line.

File: .openshift-ci/drivers/build.Dockerfile
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/build.Dockerfile:1]: The base image is replaced by Openshift CI. Robby advised me to use something like "FROM will_be_replaced_by_openshift_ci"

File: .openshift-ci/drivers/drivers.Dockerfile
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/drivers.Dockerfile:11]: Maybe move the scripts you are using from .circlecli to .openshift-ci.

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/scripts/kernel-splitter.py:20]: Maybe print something here.

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/scripts/kernel-splitter.py:53]: I don't think it is very clear what v is and it is used a lot.

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/scripts/kernel-splitter.py:97]: get_tps is not a very clear name.

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/scripts/kernel-splitter.py:128]: I feel like there is a more Pythonic way of doing this. Python has a max function, and one of the parameters can be a lambda. Might not worth doing in this PR.

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (JoukoVirtanen) [inline .openshift-ci/drivers/scripts/kernel-splitter.py:140]: Maybe we could have a dictionary of builders. That way when we a new builder we don't have to add any other code. This could be done in a future PR.

File: kernel-modules/dockerized/scripts/patch-files.sh
Comment (JoukoVirtanen) [inline kernel-modules/dockerized/scripts/patch-files.sh:3]: Remember to remove the x before merging.

File: .circleci/kernel-module-build-failures-check/20-test-for-build-failures.sh
Comment (Molter73) [reply .circleci/kernel-module-build-failures-check/20-test-for-build-failures.sh:8]: I'm not sure I follow, this change is meant to allow the directory being scanned for failures when building drivers in OSCI. This is because the builders in OSCI directly dump logs to `/FAILURES` and I didn't want to duplicate the script for this small change. How would `find` help in this situation?

File: .openshift-ci/drivers/build.Dockerfile
Comment (Molter73) [reply .openshift-ci/drivers/build.Dockerfile:1]: I honestly don't like having that in a Dockerfile, can we open the discussion up to the rest of the team and see what everyone thinks?

File: .openshift-ci/drivers/drivers.Dockerfile
Comment (Molter73) [reply .openshift-ci/drivers/drivers.Dockerfile:11]: If I do that, I will need to also modify `.circleci/config.yaml` or duplicate the script, I think it would be cleaner to do so on a follow up PR once we completely remove CircleCI, WDYT?

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (Molter73) [reply .openshift-ci/drivers/scripts/kernel-splitter.py:97]: I'll rename it to `get_tasks_per_shard`

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (Molter73) [reply .openshift-ci/drivers/scripts/kernel-splitter.py:128]: I'll try and change it for a call to `min(builders, key=lambda b: b.get_tasks_per_shard())`

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (Molter73) [reply .openshift-ci/drivers/scripts/kernel-splitter.py:132]: Done

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (Molter73) [reply .openshift-ci/drivers/scripts/kernel-splitter.py:20]: I've print a message if `Task.parse` fails, that way I keep the responsibility of reporting separated from parsing the line, does that work?

File: .openshift-ci/drivers/BLOCKLIST
Comment (Molter73) [reply .openshift-ci/drivers/BLOCKLIST:9]: Done

File: .openshift-ci/drivers/drivers.Dockerfile
Comment (JoukoVirtanen) [reply .openshift-ci/drivers/drivers.Dockerfile:11]: Makes sense.

File: .openshift-ci/drivers/build.Dockerfile
Comment (JoukoVirtanen) [reply .openshift-ci/drivers/build.Dockerfile:1]: I posted about this in slack. Seems like Gavin and Dmitrii like the idea of having a placeholder base. You can keep this as is for now.

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (JoukoVirtanen) [reply .openshift-ci/drivers/scripts/kernel-splitter.py:20]: Yes

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (JoukoVirtanen) [reply .openshift-ci/drivers/scripts/kernel-splitter.py:132]: You could pass that list to distribute_shards and use the list for dump().

**Comment by JoukoVirtanen** [comment]
Rebasing would probably be a good idea.

File: .openshift-ci/drivers/scripts/kernel-splitter.py
Comment (Molter73) [reply .openshift-ci/drivers/scripts/kernel-splitter.py:131]: Also done 🙂

**Review by robbycochran — APPROVED** [review]

**Comment by Molter73** [comment]
I'll merge this as soon as OSCI finishes running (had to rebase on a minor conflict). Any errors/problems/unaddressed changes I'll make on a follow up.

**Pull request merged**
